scapy/contrib/automotive/gm/gmlan.py

```diff
@@ -0,0 +1,644 @@
+        0xD3: "$D3: SoftwareModuleIdentifierAlphaCode (SWMIAC) 03",
+        0xD4: "$D4: SoftwareModuleIdentifierAlphaCode (SWMIAC) 04",
+        0xD5: "$D5: SoftwareModuleIdentifierAlphaCode (SWMIAC) 05",
+        0xD6: "$D6: SoftwareModuleIdentifierAlphaCode (SWMIAC) 06",
+        0xD7: "$D7: SoftwareModuleIdentifierAlphaCode (SWMIAC) 07",
+        0xD8: "$D8: SoftwareModuleIdentifierAlphaCode (SWMIAC) 08",
+        0xD9: "$D9: SoftwareModuleIdentifierAlphaCode (SWMIAC) 09",
+        0xDA: "$DA: SoftwareModuleIdentifierAlphaCode (SWMIAC) 10",
+        0xDB: "$DB: EndModelPartNumberAlphaCode",
+        0xDC: "$DC: BaseModelPartNumberAlphaCode",
+        0xDD: "$DD: SoftwareModuleIdentifierDataIdentifiers (SWMIDID)",
+        0xDE: "$DE: GMLANIdentificationData (GMLANID)",
+        0xDF: "$DF: ECUOdometerValue (ECUODO)",
+        0xE0: "$E0: VehicleLevelDataRecord (VLDR) 0",
+        0xE1: "$E1: VehicleLevelDataRecord (VLDR) 1",
+        0xE2: "$E2: VehicleLevelDataRecord (VLDR) 2",
+        0xE3: "$E3: VehicleLevelDataRecord (VLDR) 3",
+        0xE4: "$E4: VehicleLevelDataRecord (VLDR) 4",
+        0xE5: "$E5: VehicleLevelDataRecord (VLDR) 5",
+        0xE6: "$E6: VehicleLevelDataRecord (VLDR) 6",
+        0xE7: "$E7: VehicleLevelDataRecord (VLDR) 7",
+        0xE8: "$E8: Subnet_Config_List_GMLANPowertrainExpansionBus (SCLGPEB)",
+        0xE9: "$E9: Subnet_Config_List_GMLANFrontObjectExpansionBus "
+              "(SCLGFOEB)",
+        0xEA: "$EA: Subnet_Config_List_GMLANRearObjectExpansionBus (SCLGROEB)",
+        0xEB: "$EB: Subnet_Config_List_GMLANExpansionBus1 (SCLGEB1)",
+        0xEC: "$EC: Subnet_Config_List_GMLANExpansionBus2 (SCLGEB2)",
+        0xED: "$ED: Subnet_Config_List_GMLANExpansionBus3 (SCLGEB3)",
+        0xEE: "$EE: Subnet_Config_List_GMLANExpansionBus4 (SCLGEB4)",
+        0xEF: "$EF: Subnet_Config_List_GMLANExpansionBus5 (SCLGEB5)",
+    })
+
+    name = 'ReadDataByIdentifier'
+    fields_desc = [
+        XByteEnumField('dataIdentifier', 0, dataIdentifiers)
+    ]
+
+
+bind_layers(GMLAN, GMLAN_RDBI, service=0x1A)
+
+
+class GMLAN_RDBIPR(Packet):
+    name = 'ReadDataByIdentifierPositiveResponse'
+    fields_desc = [
+        XByteEnumField('dataIdentifier', 0, GMLAN_RDBI.dataIdentifiers),
+        StrField('dataRecord', b'\x00')
+    ]
+
+
+bind_layers(GMLAN, GMLAN_RDBIPR, service=0x5A)
+
+
+# ########################RDBI###################################
+class GMLAN_RDBPI(Packet):
+    dataIdentifiers = ObservableDict({
+        0x0005: "OBD_EngineCoolantTemperature",
+        0x000C: "OBD_EngineRPM",
+        0x001f: "OBD_TimeSinceEngineStart"
+    })
+    name = 'ReadDataByParameterIdentifier'
+    fields_desc = [
+        FieldListField("identifiers", [],
+                       XShortEnumField('parameterIdentifier', 0,
+                                       dataIdentifiers))
+    ]
+
+
+bind_layers(GMLAN, GMLAN_RDBPI, service=0x22)
+
+
+# TODO: Implement multiple response records (length is not determinable)
```

Comment:
What is the issue here ? Could this be fixes using `PacketListField` or `FieldListField` ?

Reply:
The response is dependant on the identifier. Since this GMLAN protocol is implemented in the most generic way, I don't have any knowledge about the possible identifiers. This identifiers are dependant on the ECU, again. So without further documentation its not possible to make a decision. This packet here is just a first layer. The plan is, to extend this packet with uniq packet definitions for every identifier at runtime. Therefor I made this observableDict, that one is able to modify the protocol based on the current ECU at runtime.